### PR TITLE
Add Sharepoint Extractor

### PR DESCRIPTION
**Pull request by rishimo**
### 🤔 Why?

[SC23114](https://app.shortcut.com/metaphor-data/story/23114/create-initial-extractor-for-sharepoint)

Adds a connector to retrieve documents from Sharepoint instance(s).

### 🤓 What?

Adds extractor, helper function, test files, and documentation.

### 🧪 Tested?

Extractor works via CLI and generates valid, searchable documents. Will set up connector action for triggering and scheduling as well.

### ☑️ Checks

- [x] My PR contains actual code changes, and I have updated the version number in `pyproject.toml`.

### CODE REVIEW

File: metaphor/sharepoint/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+azure_openAI_key: <azure_openAI_key>
+
+azure_openAI_endpoint: <azure_openAI_endpoint>
+```
+
+### Optional Configurations
+
+These defaults are provided; you don't have to manually configure them.
+
+`include_text` refers to if you'd like to include the original document text alongside the embedded content.
```
Comment (prith189) [inline metaphor/sharepoint/README.md:33]: maybe use same description as [here](https://github.com/MetaphorData/connectors/blob/5944033ff8e760f859634032b5230121c5de54dd/metaphor/confluence/README.md?plain=1#L69)?

File: metaphor/sharepoint/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+```yaml
+azure_openAI_version: <azure_openAI_version> # "2023-12-01-preview"
+azure_openAI_model_name: <azure_openAI_model_name> # "Embedding_ada002"
```
Comment (prith189) [inline metaphor/sharepoint/README.md:37]: use updated embedding model in comments?

File: metaphor/sharepoint/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+azure_openAI_model: <azure_openAI_model> # "text-embedding-ada-002"
```
Comment (prith189) [inline metaphor/sharepoint/README.md:38]: use updated embedding model in comments?

File: metaphor/sharepoint/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+### Optional Configurations
+
+These defaults are provided; you don't have to manually configure them.
+
+`include_text` refers to if you'd like to include the original document text alongside the embedded content.
+
+```yaml
+azure_openAI_version: <azure_openAI_version> # "2023-12-01-preview"
```
Comment (prith189) [inline metaphor/sharepoint/README.md:36]: use updated api version?

File: tests/sharepoint/test_extractor.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+dummy_config = SharepointRunConfig(
+    sharepoint_client_id="client_id",
+    sharepoint_client_secret="client_secret",
+    sharepoint_tenant_id="tenant_id",
+    azure_openAI_key="azure_openAI_key",
+    azure_openAI_version="azure_openAI_version",
+    azure_openAI_endpoint="azure_openAI_endpoint",
+    azure_openAI_model="text-embedding-3-small",
+    azure_openAI_model_name="Embedding_3_small",
+    include_text=True,
```
Comment (prith189) [inline tests/sharepoint/test_extractor.py:20]: possible to add a test with include_text=False?

**Comment by rishimo** [comment]
`databricks-sql-connector` v3.1.2 introduces a bunch of typechecking errors, so I fixed the version as 3.1.1 until this can be remedied later.

```
metaphor/unity_catalog/utils.py:72: error: List comprehension has incompatible type List[Optional[str]]; expected List[int]  [misc]
metaphor/unity_catalog/extractor.py:224: error: Argument 1 to "_get_source_url" of "UnityCatalogExtractor" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/extractor.py:224: error: Argument 2 to "_get_source_url" of "UnityCatalogExtractor" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/extractor.py:224: error: Argument 3 to "_get_source_url" of "UnityCatalogExtractor" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/extractor.py:227: error: Argument 1 to "from_timestamp_ms" has incompatible type "Optional[int]"; expected "int"  [arg-type]
metaphor/unity_catalog/extractor.py:228: error: Argument 1 to "from_timestamp_ms" has incompatible type "Optional[int]"; expected "int"  [arg-type]
metaphor/unity_catalog/extractor.py:346: error: Argument 1 to "user_id_or_email" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/extractor.py:356: error: Item "None" of "Optional[str]" has no attribute "encode"  [union-attr]
metaphor/unity_catalog/profile/extractor.py:77: error: Argument 1 to "include_database" of "DatasetFilter" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:82: error: Argument 1 to "list" of "SchemasAPI" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:83: error: Argument 1 to "include_schema" of "DatasetFilter" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:83: error: Argument 2 to "include_schema" of "DatasetFilter" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:88: error: Argument 1 to "list" of "TablesAPI" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:88: error: Argument 2 to "list" of "TablesAPI" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:89: error: Argument 1 to "include_table" of "DatasetFilter" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:89: error: Argument 2 to "include_table" of "DatasetFilter" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:89: error: Argument 3 to "include_table" of "DatasetFilter" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:106: error: Argument 1 to "normalize_full_dataset_name" has incompatible type "Optional[str]"; expected "str"  [arg-type]
metaphor/unity_catalog/profile/extractor.py:122: error: Argument 1 to "escape_special_characters" has incompatible type "Optional[str]"; expected "str"  [arg-type]
Found 19 errors in 3 files (checked 378 source files)
```

**Review by prith189 — APPROVED** [review]

**Pull request merged**